python/hail/tests/test_typecheck2.py

```diff
@@ -0,0 +1,367 @@
+            f_list(tuple([1, 2, 3]))
+        with self.assertRaises(TypeError):
+            f_list_int(['1'])
+        with self.assertRaises(TypeError):
+            f_list_int(tuple([]))
+        with self.assertRaises(TypeError):
+            f_list_int(tuple([1]))
+
+    def test_sequence(self):
+        def f_sequence(x: Sequence):
+            typecheck(f_sequence)
+
+        def f_sequence_float(x: Sequence[float]):
+            typecheck(f_sequence_float)
+
+        f_sequence([])
+        f_sequence(tuple())
+
+        f_sequence_float([])
+        f_sequence_float(tuple())
+        f_sequence_float([1, 1.5, 2])
+        f_sequence_float((1, 1.5, 2))
+
+        with self.assertRaises(TypeError):
+            f_sequence({})
+        with self.assertRaises(TypeError):
+            f_sequence(set([1, 2, 3]))
+        with self.assertRaises(TypeError):
+            f_sequence_float(['1'])
+        with self.assertRaises(TypeError):
+            f_sequence_float(set())
+        with self.assertRaises(TypeError):
+            f_sequence_float(tuple(['1']))
+
+    def test_set(self):
+        def f_set(x: Set):
+            typecheck(f_set)
+
+        def f_set_union(x: Set[Union[int, str]]):
+            typecheck(f_set_union)
+
+        f_set(set())
+        f_set({1, 2, 3})
+        f_set({1, 'a', None})
+
+        f_set_union(set())
+        f_set_union({'1'})
+        f_set_union({'1', 1})
+        f_set_union({1, 2, 3, 4})
+
+        with self.assertRaises(TypeError):
+            f_set([])
+        with self.assertRaises(TypeError):
+            f_set(())
+        with self.assertRaises(TypeError):
+            f_set(frozenset())
+        with self.assertRaises(TypeError):
+            f_set_union({1, 1.5})
+
+    def test_frozenset(self):
+        def f_frozenset(x: FrozenSet):
+            typecheck(f_frozenset)
+
+        def f_frozenset_int(x: FrozenSet[int]):
+            typecheck(f_frozenset_int)
+
+        f_frozenset(frozenset())
+        f_frozenset(frozenset([1, 2, 3]))
+
+        f_frozenset_int(frozenset())
+        f_frozenset_int(frozenset([1, 2, 3]))
+
+        with self.assertRaises(TypeError):
+            f_frozenset(set())
+        with self.assertRaises(TypeError):
+            f_frozenset([])
+        with self.assertRaises(TypeError):
+            f_frozenset_int(set())
+        with self.assertRaises(TypeError):
+            f_frozenset_int(frozenset(['1']))
+
+    def test_collection(self):
+        def f_collection(x: Collection):
+            typecheck(f_collection)
+
+        def f_collection_str(x: Collection[str]):
+            typecheck(f_collection_str)
+
+        f_collection([])
+        f_collection([x for x in []])
+        f_collection([1, 2, 3])
+        f_collection({1, 2, 3})
+        f_collection({})
+        f_collection({}.keys())
+        f_collection({}.items())
+
+        f_collection_str([1, 2, 3])
+        f_collection_str(tuple(x for x in [1, 2, 3]))
```

Comment:
Why do these pass?

Reply:
huh, good catch. will investigate

Reply:
ah, I changed `Iterable` to `Collection` at one point. It's not possible to typecheck iterables in this framework, since iterables have no guarantee of being able to iterate twice.